[14.0][IMP] cetmix_tower_server: file create command update

Kurama2d opened the pull request:
Add `if_file_exists` and `disconnect_file` fields to "Create file using template" command. 
Supports skipping, overwriting, or raising an error when uploading a file that already exists.

Task: 4740


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added options to control file creation behavior when a file already exists on the server: skip, overwrite, or raise an error.
  - Introduced a setting to disconnect created files from their templates after command execution.
- **Bug Fixes**
  - Enhanced file-template linkage handling based on new command settings.
- **Tests**
  - Added tests covering file creation behaviors and template disconnection.
- **User Interface**
  - Updated command form to include new fields for file existence handling and template disconnection, shown for applicable actions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -127,7 +127,9 @@ def action_open_files(self):
-    def create_file(self, server, server_dir="", raise_if_exists=False):
+    def create_file(
+        self, server, if_file_exists="raise", server_dir="", raise_if_exists=False
+    ):
```
I prefer to keep order that had been here before my changes and only wanted to add new argument.

Kurama2d replied:
I removed raise_if_exists param

Kurama2d commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1199,6 +1201,7 @@ def _command_runner_file_using_template(
+                if_file_exists=command.if_file_exists,
                 server_dir=server_dir,
                 raise_if_exists=True,
             )
```
We need raise_if_exists=True for the case when an error raises when if_file_exists is "raise" and raise_if_exists=True.

Kurama2d replied:
You was right, raise_if_exists is not needed anymore

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -127,7 +127,7 @@ def action_open_files(self):
 
     # -- Business logic
-    def create_file(self, server, server_dir="", raise_if_exists=False):
+    def create_file(self, server, if_file_exists="raise", server_dir=""):
```
It's not really needed here.

Kurama2d commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,21 @@ class CxTowerCommand(models.Model):
+        "- Overwrite: Replace existing file with new one.\n"
+        "- Raise Error: Raise an error if the file already exists.",
+    )
+    is_disconnect_from_tmpl = fields.Boolean(string="Disconnect from Template")
```
Thanks, help is added

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -178,3 +179,5 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
             }
         )
+        if not existing_file or (existing_file and if_file_exists == "overwrite"):
+            return new_file
```
Thanks, made this change

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,58 @@ def test_run_command_no_command_log(self):
+        command.is_disconnect_from_tmpl = True
+        self.server_test_1.run_command(command=command)
+        template_id = self.server_test_1.file_ids.template_id
+        self.assertFalse(template_id, "File must be disconnected from template")
```
Thanks, changed this logic

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,58 @@ def test_run_command_no_command_log(self):
+            file,
+            "File must be overwritten on the server",
+        )
+
```
We need to write skip first to create new file, because we have default value 'raise' for this command and if there are any file already on the server to start with this case or if there is no file on the server just create new one and check 'skip' case

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,24 @@ class CxTowerCommand(models.Model):
+        default="skip",
+        help="What to do if file already exists on the server.\n"
+        "- Skip: Do not create or update the file.\n"
+        "- Overwrite: Replace existing file with new one.\n"
```
"- Overwrite: Replace the existing the file with the new one.\n"

Kurama2d replied:
Thanks, added articles to their positions

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,24 @@ class CxTowerCommand(models.Model):
+    )
+    is_disconnect_from_tmpl = fields.Boolean(
+        string="Disconnect from Template",
+        help="If enabled, disconnect from the template after running the command.\n",
```
help="If enabled, disconnect from the file template after running the command.\n",

Kurama2d replied:
Thanks, changed a little

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -127,7 +127,7 @@ def action_open_files(self):
 
     # -- Business logic
-    def create_file(self, server, server_dir="", raise_if_exists=False):
+    def create_file(self, server, if_file_exists="raise", server_dir=""):
```
Restrain from switching function arguments position as it may lead to an unpredictable behaviour in the modules that inherit this one.

Kurama2d replied:
Thanks, moved if_file_exists after server_dir to avoid confusion.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1232,6 +1234,9 @@ def _command_runner_file_using_template(
                 )
 
+            if command.is_disconnect_from_tmpl:
+                file.template_id = False
```
Use [the existing function](https://github.com/cetmix/cetmix-tower/blob/ae60ad81c1da9554ecbfb849a2717ef3191c1009/cetmix_tower_server/models/cx_tower_file.py#L314) for that to preserve any custom logic it may implement.

Kurama2d replied:
Thanks, replaced to this func

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,27 @@ class CxTowerCommand(models.Model):
+            ("overwrite", "Overwrite"),
+            ("raise", "Raise Error"),
+        ],
+        string="If File Exists",
```
No need to add a string here explicitly, it will be [generated automatically](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#fields) from the field name.

Kurama2d replied:
Yes, you are right. Removed the string.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1165,6 +1166,7 @@ def _command_runner(
     def _command_runner_file_using_template(
         self,
         log_record,
+        command,
```
Do we actually need the command? We can always get if from the `plan_log` record. 
Meanwhile we can drop `file_template` argument either  as we can also get if from the command later in the runner function.

Kurama2d replied:
Double checked, we don't really need it. Also removed file_template argument and replace it to call template from log and command directly.

ivs-cetmix commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,76 @@ def test_run_command_no_command_log(self):
+        command.is_disconnect_from_tmpl = True
+        self.server_test_1.run_command(command=command)
+
+        new_files = self.server_test_1.file_ids - initial_files
```
We need to ensure that the `new_files` is exactly the file that was uploaded with the command. Just to avoid any potential issues.

Kurama2d replied:
Here in this case the only way to check that new_files is exactly the file that was uploaded with the command if the assert with empty template. We don't have template as the source of date to check connection with command after command run. We don't have any flag to ensure in command log, command or server.

ivs-cetmix replied:
Agree, however we can check:
a) that there is only one file was created `len(new_files) == 1`
b) that that file has the same code as the template `file.code == template.code`

Kurama2d replied:
Added 2 asserts, but for file `code_on_server` field value is taken from the template.

ivs-cetmix reviewed (changes requested):
Some minor comments, otherwise looks good.

GabbasovDinar reviewed (approved):
LGTM

ivs-cetmix reviewed (changes requested):
Please add this new field into the list of exported fields in the [cetmix_tower_yaml](https://github.com/cetmix/cetmix-tower/blob/f5787bee25540a8f9c9269863db1d352a0a45d8d/cetmix_tower_yaml/models/cx_tower_command.py#L10) module. This should be done in  a separate commit.

Kurama2d commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,26 @@ class CxTowerCommand(models.Model):
         column2="variable_id",
     )
 
+    if_file_exists = fields.Selection(
+        selection=[
+            ("skip", "Skip"),
+            ("overwrite", "Overwrite"),
+            ("raise", "Raise Error"),
+        ],
+        default="skip",
+        help="What to do if file already exists on the server.\n"
+        "- Skip: Do not create or update the file.\n"
+        "- Overwrite: Replace the existing file with the new one.\n"
+        "- Raise Error: Raise an error if the file already exists.",
+    )
+    is_disconnect_from_tmpl = fields.Boolean(
+        string="Disconnect from Template",
+        help=(
+            "If enabled, disconnects the file from its template "
+            "after running the command.\n"
+        ),
+    )
```
In the create_file if_file_exists default value is raise to save previous behavior, but in command itself default value should be skip.

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,82 @@ def test_run_command_no_command_log(self):
+                "server_ids": [(4, self.server_test_1.id)],
+                "command_ids": [(4, command.id)],
+            }
+        )
```
Thanks. removed

Kurama2d commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -140,6 +140,26 @@ class CxTowerCommand(models.Model):
+        help=(
+            "If enabled, disconnects the file from its template "
+            "after running the command.\n",
+        ),
```
Thanks. fixed

ivs-cetmix reviewed (changes requested):
Looks good, however this is still pending:

Please add this new field into the list of exported fields in the [cetmix_tower_yaml](https://github.com/cetmix/cetmix-tower/blob/f5787bee25540a8f9c9269863db1d352a0a45d8d/cetmix_tower_yaml/models/cx_tower_command.py#L10) module. This should be done in a separate commit.

ivs-cetmix commented:
@Kurama2d please fix the PR name. The mandatory version tag is missing.

Kurama2d commented on cetmix_tower_yaml/tests/test_command.py:
```diff
@@ -162,6 +174,16 @@ def test_yaml(command):
                 command.reference,
                 self.command_test_yaml_dict["reference"],
                 "YAML value doesn't match Cetmix Tower one",
+            )            
```
Fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -179,5 +179,7 @@ def create_file(self, server, server_dir="", if_file_exists="raise"):
             }
         )
+        if existing_file and if_file_exists == "overwrite":
+            existing_file.unlink()
```
This may [physically delete](https://github.com/cetmix/cetmix-tower/blob/c653f3bae1e144130c6a741cd5279e0b3b3a89c8/cetmix_tower_server/models/cx_tower_file.py#L299) file from server. Is this what we really want to do?

Kurama2d replied:
It's one of the options, but I decided to change overwrite logic to overwrite data of the existing file.

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -162,12 +162,13 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
-            if raise_if_exists:
-                raise ValidationError(_("File already exists on server."))
+        if existing_file and if_file_exists == "skip":
             return existing_file
 
-        return file_model.with_context(is_custom_server_dir=True).create(
+        if existing_file and if_file_exists == "raise":
+            raise ValidationError(_("File already exists on server."))
+
```
I think it's better to have separate conditions for more clear understanding

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,86 @@ def test_run_command_no_command_log(self):
+        command.is_disconnect_from_tmpl = True
+        self.server_test_1.run_command(command=command)
+
+        new_files = self.server_test_1.file_ids - initial_files
+        self.assertEqual(len(new_files), 1, "Must be one new file created")
+        self.assertEqual(
+            new_files.code_on_server,
+            command.file_template_id.code,
+            "File code must match template",
+        )
+        self.assertFalse(
+            new_files.template_id, "File must be disconnected from template"
+        )
```
It's extra, we have assert for files count above

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,86 @@ def test_run_command_no_command_log(self):
+            file_template.source,
+            "File source must match template after overwrite",
+        )
+
```
Added

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,105 @@ def test_run_command_no_command_log(self):
+            ),
+            1,
+            "There must be exactly one physical file record after skip",
+        )
```
Put this assert to it's place after skip part

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,105 @@ def test_run_command_no_command_log(self):
+            "File source must match template after overwrite",
+        )
+
+    def test_is_file_disconnected_from_template(self):
+        """Test if file is disconnected from template after being created."""
+
+        initial_files = self.server_test_1.file_ids
+        command = self.command_create_file_with_template_server_source
+
+        command.is_disconnect_from_tmpl = True
+        self.server_test_1.run_command(command=command)
+
+        new_files = self.server_test_1.file_ids - initial_files
+        self.assertEqual(len(new_files), 1, "Must be one new file created")
+        self.assertEqual(
+            new_files.code_on_server,
+            command.file_template_id.code,
+            "File code must match template",
+        )
+        self.assertFalse(
+            new_files.template_id, "File must be disconnected from template"
+        )
```
It's not needed here

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -162,12 +162,27 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+            existing_file.with_context(is_custom_server_dir=True).write(
+                {
+                    "name": self.file_name,
+                    "code_on_server": self.code,
```
It's a helper field that is used for manual checks only.
You should use existing functions instead of doing direct writes and manual updates. There is an [example](https://github.com/cetmix/cetmix-tower/blob/0075aa48fe89112d371b9b445f990e588c98f8ac/cetmix_tower_server/models/cx_tower_file_template.py#L115) in the same file.

Kurama2d replied:
Thanks for noticing, replaced with method

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -162,12 +162,21 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+                existing_file._get_file_values_from_related_template()
+            )
+            if existing_file.source == "tower":
+                existing_file.action_push_to_server()
```
Please check the [write](https://github.com/cetmix/cetmix-tower/blob/70e6b9e72ae46431bed8a90495b294352de55d47/cetmix_tower_server/models/cx_tower_file.py#L290) method. The file will be synced automatically if `auto_sync` is enabled, no need to sync it twice.

Kurama2d replied:
Thanks, removed push to the server

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,105 @@ def test_run_command_no_command_log(self):
+            orig_file,
+            skipped_file,
+            "Skip should return the existing file, not create a new one",
+        )
```
It's not needed to compare id as we compare records

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -162,12 +162,19 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
-            if raise_if_exists:
-                raise ValidationError(_("File already exists on server."))
+        if existing_file and if_file_exists == "skip":
             return existing_file
 
-        return file_model.with_context(is_custom_server_dir=True).create(
+        if existing_file and if_file_exists == "raise":
+            raise ValidationError(_("File already exists on server."))
+
+        if existing_file and if_file_exists == "overwrite":
+            existing_file.with_context(is_custom_server_dir=True).write(
+                existing_file._get_file_values_from_related_template()
+            )
+            return existing_file
+
```
I think it's extra because we can catch error in _get_file_values_from_related_template

ivs-cetmix replied:
@Kurama2d makes sense, could you please check this?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -162,12 +162,19 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+
+        if existing_file and if_file_exists == "overwrite":
+            existing_file.with_context(is_custom_server_dir=True).write(
+                existing_file._get_file_values_from_related_template()
```
What if existing file is not linked to the same file template or it's not linked to any template at all?
Meanwhile [here](https://github.com/cetmix/cetmix-tower/blob/37b03fb78f9edeea1d93d5f989e954ebf4240785/cetmix_tower_server/models/cx_tower_file_template.py#L158) you are searching for a template linked file only, while not taking into account the case mentioned above.

Kurama2d replied:
Thanks, I covered the cases too

ivs-cetmix reviewed (changes requested):
NB: plese do NOT squash commits until you get the final approval.

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,109 @@ def test_run_command_no_command_log(self):
+                if_file_exists=command.if_file_exists,
+            )
+        # Change command to "overwrite" file if it exists
+        command.write({"if_file_exists": "overwrite"})
+        # Run command again, it should overwrite the file
+        file_template.create_file(
+            server=self.server_test_1,
+            server_dir=file_template.server_dir,
+            if_file_exists=command.if_file_exists,
+        )
```
Below we check that there is only 1 file oh the server, it's the same

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,66 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+            If the file already exists on the server if_file_exists is set to "raise".
         """
         self.ensure_one()
+        # Explicit guard against invalid behavior values
+        valid_behaviors = {"skip", "raise", "overwrite"}
+        if if_file_exists not in valid_behaviors:
+            raise ValidationError(
+                f"Invalid if_file_exists value: {if_file_exists}. "
+                f"Expected one of {valid_behaviors}."
+            )
```
It's not neccessary here

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,66 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        existing_file = (
+            existing_file_ids.filtered(lambda f: f.template_id == self)[:1]
+            or existing_file_ids[:1]
+        )
```
Thanks, added order but not limit because we have template check below

Kurama2d commented on cetmix_tower_server/tests/test_file.py:
```diff
@@ -302,18 +302,28 @@ def test_create_file_with_template(self):
         )
 
-        file = file_template.create_file(server=self.server_test_1)
+        file = file_template.create_file(
+            server=self.server_test_1,
+            server_dir=file_template.server_dir,
+            if_file_exists="overwrite",
+        )
         self.assertEqual(file.code, self.file_template.code)
         self.assertEqual(file.template_id, file_template)
         self.assertEqual(file.server_id, self.server_test_1)
         self.assertEqual(file.source, "tower")
         self.assertEqual(file.server_dir, self.file_template.server_dir)
 
```
It's not my test so I prefer not to touch anything except something related to my changes

Kurama2d commented on cetmix_tower_server/tests/test_file.py:
```diff
@@ -302,18 +302,28 @@ def test_create_file_with_template(self):
         with self.assertRaises(exceptions.ValidationError):
-            file_template.create_file(server=self.server_test_1, raise_if_exists=True)
+            file_template.create_file(
+                server=self.server_test_1,
+                server_dir=file_template.server_dir,
+                if_file_exists="raise",
+            )
 
         another_file = file_template.create_file(
-            server=self.server_test_1, raise_if_exists=False
+            server=self.server_test_1,
+            server_dir=file_template.server_dir,
+            if_file_exists="skip",
         )
```
Not really needed here

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,38 +136,55 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        existing_file = (
+            existing_file_ids.filtered(lambda f: f.template_id == self)[:1]
+            or existing_file_ids[:1]
+        )
```
Added

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,38 +136,55 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        if existing_file and if_file_exists == "overwrite":
+            existing_file.with_context(is_custom_server_dir=True).write(
+                existing_file._get_file_values_from_related_template()
+            )
```
Yes, replaced to vals

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,67 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+            existing_file_ids.filtered(lambda f: f.template_id == self)[:1]
+            or existing_file_ids[:1]
+        )
+        if existing_file and if_file_exists == "skip":
+            return existing_file
+
+        if existing_file and if_file_exists == "raise":
+            raise ValidationError(_("File already exists on server."))
+
```
I keep the previous behavir of the method

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,110 @@ def test_run_command_no_command_log(self):
+    def test_is_file_disconnected_from_template(self):
+        """Test if file is disconnected from template after being created."""
+
+        initial_files = self.server_test_1.file_ids
+        command = self.command_create_file_with_template_server_source
+
+        command.is_disconnect_from_tmpl = True
+        self.server_test_1.run_command(command=command)
+
+        new_files = self.server_test_1.file_ids - initial_files
+        self.assertEqual(len(new_files), 1, "Must be one new file created")
+        self.assertEqual(
```
It's not really needed here

Kurama2d commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -1265,6 +1265,109 @@ def test_run_command_no_command_log(self):
+            skipped_file,
+            "Skip should return the existing file, not create a new one",
+        )
+        self.assertEqual(
+            self.env["cx.tower.file"].search_count(
+                [
+                    ("template_id", "=", file_template.id),
+                    ("server_id", "=", self.server_test_1.id),
+                ]
+            ),
+            1,
+            "There must be exactly one physical file record after skip",
+        )
```
Added

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -152,38 +152,51 @@ def create_file(self, server, server_dir="", if_file_exists="raise"):
             [
-                ("template_id", "=", self.id),
                 ("server_id", "=", server.id),
                 ("server_dir", "=", server_dir or self.server_dir),
```
This will return files all files with the same source that are located in the same directory.
We need to check the `name` field as well to avoid it.

Kurama2d replied:
Thanks, added `name` field to domain

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -152,38 +152,51 @@ def create_file(self, server, server_dir="", if_file_exists="raise"):
+            order="id DESC",
+        )
+        # Filter existing files by template if it exists, otherwise take the first one
+        existing_file = (
```
Why should we do an additional sorting if all the files are already retrieved with the query?
Cannot we simply return `existing_file_ids and existing_file_ids[0]`?

Kurama2d replied:
Because in previous logic we searched the file only with particular template. So to save this logic and combine it with cases when there are no template at all I check if there any file with curent template and if now we can take any file

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -152,38 +152,51 @@ def create_file(self, server, server_dir="", if_file_exists="raise"):
+        vals = {
+            "template_id": self.id,
+            "name": self.file_name,
+            "code_on_server": self.code,
```
Why are you modifying [this field](https://github.com/cetmix/cetmix-tower/blob/7a658a7f3093208aa9c4818047a3a1a3d22fdc13/cetmix_tower_server/models/cx_tower_file.py#L111) instead of `code`?

Kurama2d replied:
Thanks for noticing, it's my mistake, changed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,67 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+
+        vals = {
+            "template_id": self.id,
+            "name": self.file_name,
```
If the name is the same, does it make any sense to overwrite it?

Kurama2d replied:
After adding to the search domain it doesn't make sense, removed from overwriting

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,68 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        existing_file_ids = file_model.search(
             [
-                ("template_id", "=", self.id),
+                ("name", "=", self.file_name),
```
Here is another case we need to keep in mind: "name" field can contain variables. So a file with the name value `/opt/test` and another file with the name value `/opt/{{ branch_name}}` rendered with variable value `branch="test"` will end up with the same file but different file names.

Kurama2d replied:
I added another filtered for name and rendered_name, because rendered_name is not storable field and we can't use it in search and there no better decisions here unfortunately.

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        :param if_file_exists: str, optional
+            Defines the behavior if the file already exists on the server.
         :param server_dir: str, optional
             The directory on the server where the file should be created. If not set,
```
All lines lenght checked by pre-commiter

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        # Filter existing files by name or rendered name
+        existing_file_ids = existing_file_ids.filtered(
+            lambda f: f.name in self.file_name or f.rendered_name in self.file_name
+        )
```
You are right, fixed it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+            raise ValidationError(_("File already exists on server."))
+
+        vals = {
+            "template_id": self.id,
```
You can simply add the `template_id` and then call the [onchange method](https://github.com/cetmix/cetmix-tower/blob/aa623db51b2bc44e2d0cefeadd55246ef870b437/cetmix_tower_server/models/cx_tower_file.py#L247). 
This is a better approach because it allows to ensure that no further possible changes will break the code.

Kurama2d replied:
I tried this method and many tests failed for some reason (not only mine ones). I checked data that is set in this method and it's the same as in my logic, but unfortunately something happens. I'm not seure if it worth further checking and changes here.

ivs-cetmix replied:
I think you need to check what exactly fails and why. Probably there are some minor issues that can be easily tweaked. Anyway it's not a good idea to have the same operations spread across several functions in the code. It makes the code error prone and less maintainable.

Kurama2d replied:
Changed to simply write template_id, then _inverse_template_id triggers and updates data (not onchange).

Kurama2d commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+        if existing_file and if_file_exists == "raise":
+            raise ValidationError(_("File already exists on server."))
+
+        if existing_file and if_file_exists == "overwrite":
+            existing_file.with_context(is_custom_server_dir=True).write(
+                {
+                    "template_id": self.id,
+                }
+            )
+            return existing_file
```
template_id updates and inverse triggers so we sync data from template to the file.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
             [
-                ("template_id", "=", self.id),
                 ("server_id", "=", server.id),
                 ("server_dir", "=", server_dir or self.server_dir),
```
`server_dir` is also using variables, so using it in domains may lead to an irrelevant result.

Kurama2d replied:
Reworked logic with rendered and unrendered server directory and file name

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,71 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
-                raise ValidationError(_("File already exists on server."))
+        # Filter existing files by name or rendered name
+        existing_file_ids = existing_file_ids.filtered(
+            lambda f: f.name == self.file_name or f.rendered_name == self.file_name
```
You should filter by the [full server path](https://github.com/cetmix/cetmix-tower/blob/d1fa7bb2388523bad944ac9679de1bc255b91dd1/cetmix_tower_server/models/cx_tower_file.py#L66) because it contains both render file name and directory.

Kurama2d replied:
Reworked logic with rendered and unrendered server directory and file name

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,94 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
-                "server_dir": server_dir or self.server_dir,
-                "file_type": self.file_type,
-                "source": self.source,
+                path.strip("/") if path.startswith("//") else path
```
What if path starts with `///` or some other special symbol?

I think we need to straighten the check algorithm:
1. Get the path/server_dir and the file name from the template and compose the final URL. 
2. Render them using the `render_code_custom` function like it's done [here](https://github.com/cetmix/cetmix-tower/blob/88464c5841395a3b80aa7832241e7aea2da54ead/cetmix_tower_server/models/cx_tower_file.py#L221).
3.  Compare them with the `full_server_path` values of the records got with search.

Kurama2d replied:
Agree,  changed the logic with rendering path from template

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -136,45 +136,86 @@ def create_file(self, server, server_dir="", raise_if_exists=False):
+            )
         file_model = self.env["cx.tower.file"]
-        existing_file = file_model.search(
+        existing_file_ids = file_model.search(
```
`existing_files` because we are fetching records, not ids

Kurama2d replied:
You are right, fixed

ivs-cetmix reviewed (approved):
LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

GabbasovDinar reviewed (approved):
LGTM

Kurama2d commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1232,6 +1228,17 @@ def _command_runner_file_using_template(
                     )
                 )
 
+            if log_record.command_id.is_disconnect_from_tmpl:
+                file.action_unlink_from_template()
+
+            if log_record.command_id.if_file_exists == "skip":
+                return log_record.finish(
+                    fields.Datetime.now(),
+                    0,
+                    _("File already exists on server. Upload skipped"),
+                    None,
+                )
```
No, it should be after operations. We also have is_disconnect_from_tmpl that can be set so we need to disconect the file even if it was skipped.

Aldeigja commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [14.0](https://github.com/cetmix/cetmix-tower/commits/14.0).

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -155,6 +155,26 @@ class CxTowerCommand(models.Model):
+        "- Overwrite: Replace the existing file with the new one.\n"
+        "- Raise Error: Raise an error if the file already exists.",
+    )
+    is_disconnect_from_tmpl = fields.Boolean(
```
Let's rename this field into `disconnect_file` because although `is_disconnect_from_tmpl` is compliant with OCA standards, it's not so user friendly for YAML templates.

Kurama2d replied:
Renamed.

ivs-cetmix reviewed (changes requested):
Some minor change, otherwise looks good.

Kurama2d committed:
[IMP] cetmix_tower_server: file create command update

Add `if_file_exists` and `disconnect_file` fields to "Create file using template" command.
Supports skipping, overwriting, or raising an error when uploading a file that already exists.

Task: 4740

Kurama2d commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1220,16 +1216,32 @@ def _command_runner_file_using_template(
-                    _(
-                        "File source cannot be determined: '%(source)s'",
-                        source=file.source,
+            # Context is used to detect a retry
+            # and avoid handling skip logic on first attempt
+            is_creation_skipped = file._context.get("file_creation_skipped")
+
+            if not is_creation_skipped:
+                if file.source == "server":
+                    file.action_pull_from_server()
+                elif file.source == "tower":
+                    file.action_push_to_server()
+                else:
+                    raise UserError(
+                        _(
+                            "File source cannot be determined: '%(source)s'",
+                            source=file.source,
+                        )
                     )
+
+            if log_record.command_id.disconnect_file:
+                file.action_unlink_from_template()
+
+            if is_creation_skipped:
+                return log_record.finish(
+                    fields.Datetime.now(),
+                    0,
+                    _("File already exists on server. Upload skipped"),
+                    None,
                 )
```
No changes needed anymore.

ivs-cetmix reviewed (approved):
/ocabot merge minor

CetmixGitDrone commented:
Hey, thanks for contributing! Proceeding to merge this for you.
Prepared branch [14.0-ocabot-merge-pr-309-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/14.0-ocabot-merge-pr-309-by-ivs-cetmix-bump-minor), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 130bd169f7642aeb03b13d5ade78d98884fab2a0. Thanks a lot for contributing to cetmix. ❤️
